osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SpinnerBonusComponent.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Objects.Drawables;
+
+namespace osu.Game.Rulesets.Osu.Objects.Drawables.Pieces
+{
+    /// <summary>
+    /// A component that tracks spinner spins and add bonus score for it.
+    /// </summary>
+    public class SpinnerBonusComponent : CompositeDrawable
+    {
+        private readonly DrawableSpinner drawableSpinner;
+        private readonly Container<DrawableSpinnerTick> ticks;
+        private readonly OsuSpriteText bonusCounter;
+
+        public SpinnerBonusComponent(DrawableSpinner drawableSpinner, Container<DrawableSpinnerTick> ticks)
+        {
+            this.drawableSpinner = drawableSpinner;
+            this.ticks = ticks;
+
+            drawableSpinner.OnNewResult += onNewResult;
+
+            AutoSizeAxes = Axes.Both;
+            InternalChild = bonusCounter = new OsuSpriteText
+            {
+                Anchor = Anchor.Centre,
+                Origin = Anchor.Centre,
+                Font = OsuFont.Numeric.With(size: 24),
+                Alpha = 0,
+            };
+        }
+
+        private int currentSpins;
+
+        public void UpdateRotation(double rotation)
```

Comment:
Probably rename this to `SetRotation` to avoid adding one extra dialect to setting rotation, as previously mentioned.